core/src/main/scala/scalaz/IList.scala

```diff
@@ -2,18 +2,19 @@ package scalaz
 
 import scala.annotation.tailrec
 import scala.annotation.unchecked.uncheckedVariance
-import std.option.{ cata, none, some }
-import std.stream.{ toZipper => sToZipper }
-import std.tuple.{ tuple2Bitraverse => BFT }
-import Liskov.{ <~<, refl }
-import IList.{empty, single}
+import std.option.{cata, none, some}
+import std.stream.{toZipper => sToZipper}
+import std.tuple.{tuple2Bitraverse => BFT}
+import Liskov.{<~<, refl}
+import IList.{empty, foldRightLazy, single, byNeed}
+import scalaz.std.option
 
 /**
  * Safe, invariant alternative to stdlib `List`. Most methods on `List` have a sensible equivalent
  * here, either on the `IList` interface itself or via typeclass instances (which are the same as
  * those defined for stdlib `List`). All methods are total and stack-safe.
  */
-sealed abstract class IList[A] extends Product with Serializable {
+trait IList[A] extends Product with Serializable {
```

Comment:
This was to be able to hide sub-typing between `INil`/`ICons` and `IList` to avoid any confusion when pattern matching. But maybe it is overkill. [Reverting back to abstract class](https://github.com/jbgi/scalaz/commit/231946bf5c89cac6cd3c159a21462d4615292e46) is possible.